src/main/java/org/ensime/maven/plugins/ensime/EnsimeConfigGenerator.java

```diff
@@ -295,20 +312,132 @@ private EnsimeModule ensimeProjectsToModule(final List<EnsimeProject> p) {
+          project.getDependencies();
+
+      List<org.apache.maven.model.Dependency> depMgmtDependencies = 
+          Optional.<DependencyManagement>ofNullable(
+              project.getModel().getDependencyManagement())
+          .map(depMgmt -> depMgmt.getDependencies())
+          .orElse(new ArrayList<>());
+
+      Set<Artifact> allDependencies = project.getArtifacts();
+
+      Pair<String, Optional<String>> result = getScalaVersion(directDependencies,
+          depMgmtDependencies, allDependencies, ENSIME_SCALA_VERSION, DEFAULT_SCALA_VERSION);
+
+      result._2.ifPresent(logMessage -> log.warn(logMessage));
+
+      return result._1;
+  }
+
+
+  /**
+   * Contains the pure functional logic for determining scala version.
+   *
+   * @return a pair with the first element being the determined
+   *         Scala version, and the second element being an optional
+   *         log message.
+   */
+  public static Pair<String, Optional<String>> getScalaVersion(
+      List<org.apache.maven.model.Dependency> directDependencies,
+      List<org.apache.maven.model.Dependency> depMgmtDependencies,
+      Set<Artifact> allDependencies, String ensimeScalaVersion,
+      String defaultScalaVersion) {
+
+      final String scalaLibraryGroupId = "org.scala-lang";
+      final String scalaLibraryArtifactId = "scala-library";
+
+      // Determine scala version, via a number of different methods:
+
+      // If -Densime.scala.version is set, use that
+      if (ensimeScalaVersion != null && !ensimeScalaVersion.trim().equals(""))
+          return new Pair<String, Optional<String>>(ensimeScalaVersion, Optional.empty());
+
+      // Look for a scala library dependency in <dependencies>
+      Optional<String> directVersion = directDependencies.stream()
+          .filter(d ->
+              d.getGroupId().equals(scalaLibraryGroupId) &&
+              d.getArtifactId().equals(scalaLibraryArtifactId))
+          .findFirst()
+          .map(d -> d.getVersion());
+      if (directVersion.isPresent())
+          return new Pair<String, Optional<String>>(directVersion.get(), Optional.empty());
+
+      // Look in <dependencyManagement>
+      Optional<String> depMgmtVersion = depMgmtDependencies.stream()
+          .filter(d ->
+              d.getGroupId().equals(scalaLibraryGroupId) &&
+              d.getArtifactId().equals(scalaLibraryArtifactId))
+          .findFirst()
+          .map(d -> d.getVersion());
+      if (depMgmtVersion.isPresent())
+          return new Pair<String, Optional<String>>(depMgmtVersion.get(), Optional.empty());
+
+      // Look at all dependencies, including transitive dependencies
+      String[] allVersions = allDependencies.stream()
+          .filter(d ->
+              d.getGroupId().equals(scalaLibraryGroupId) &&
+              d.getArtifactId().equals(scalaLibraryArtifactId))
+          .map(d -> d.getVersion())
+          .collect(toSet())
+          .toArray(new String[0]);
```

Comment:
I don't really understand what does this statement do? why the length of the String array is 0?

Reply:
It converts the Set into an array -- this version causes the result to be String[] instead of Object[], which is returned by Set.toArray(), since you can't easily cast Object[] to String[].  It will take the passed in array as the type of the return array.  If the passed in array is actually big enough to hold the results it will use it, if not it will create a new array and return it.  Passing in new String[0] will thus work for any set size, so you don't have to calculate the array size (especially here, where we would have to add an extra step to get the Set, then get its size, then create an array of that size).

Reply:
Mmm, thanks for the detailed explanation.